### PR TITLE
BoomPHF compatibility

### PR DESCRIPTION
**Pull request by k3yavi**
Adding BoomPHF compatibility to do super-fast single-cell mapping !

### CODE REVIEW

File: src/compression.rs [hunk 1]
```diff
@@ -557,3 +586,251 @@ pub fn compress_graph<K: Kmer, D: Clone + Debug, S: CompressionSpec<D>>(
...
+/// Compression of paths in Debruijn graph
+impl<'a, K: Kmer, D: Clone + Debug, S: CompressionSpec<D>> CompressFromHash<'a, K, D, S> {
+    fn get_kmer_data(&'a self, kmer: &K) -> (&Exts, &'a D) {
+        match self.index.get_data_for_kmer(kmer) {
+            Some(data) => data,
+            None => panic!("couldn't find kmer {:?}", kmer),
+        }
+    }
+
+    fn get_kmer_id(&self, kmer: &K) -> Result<usize, usize> {
```
Comment (pmarks) [inline src/compression.rs]: does the Err case of this Result<usize,usize> have any meaning?  If not, this should just return an Option<usize>

File: src/compression.rs [hunk 2]
```diff
@@ -557,3 +586,251 @@ pub fn compress_graph<K: Kmer, D: Clone + Debug, S: CompressionSpec<D>>(
...
+// Compress from Hash a new Struct
+//////////////////////////////
+/// Generate a compressed DeBruijn graph from hash_index
+struct CompressFromHash<'a, K: 'a + Kmer, D: 'a, S: CompressionSpec<D>> {
+    stranded: bool,
+    k: PhantomData<K>,
+    d: PhantomData<D>,
+    spec: S,
+    available_kmers: BitSet,
+    index: &'a boomphf::BoomHashMap2<K, Exts, D>,
```
Comment (pmarks) [inline src/compression.rs]: We can merge CompressFromKmers and CompressFromHash into the same implementation with the following trick: make the index field a generic type that implement a trait like this: 
```
trait KmerIndex {
    fn get_kmer_data(&'a self, kmer: &K) -> (&Exts, &'a D);
    fn get_kmer_id(&self, kmer: &K) -> Option<usize>;
}
```
then create impls of this trait for Vec<(K,Exts,D)> and BookHashMap2<K, Exts, D>.  
Then you can reuse almost all the code between CompressFromHash and CompressFromKmer.

Comment (pmarks) [reply src/compression.rs]: not sure if it's worth doing this... maybe not. We may just want to get rid of the Vec<> version.

File: src/filter.rs [hunk 3]
```diff
@@ -125,29 +163,35 @@ impl<D: Ord> KmerSummarizer<D, Vec<D>> for CountFilterSet<D> {
...
 #[inline(never)]
 pub fn filter_kmers<K: Kmer, V: Vmer<K>, D1: Clone, DS, S: KmerSummarizer<D1, DS>>(
     seqs: &[(V, Exts, D1)],
     summarizer: S,
     stranded: bool,
-) -> (Vec<(K, (Exts, DS))>, Vec<K>) {
+    report_all_kmers: bool,
+    memory_size: usize,
+)  -> ( boomphf::BoomHashMap2<K, Exts, DS>, Vec<K> )
+where DS: Debug{
```
Comment (pmarks) [inline src/filter.rs]: We shouldn't need the Debug constraint.  (BoomHashMap shouldn't need it)

File: src/filter.rs [hunk 4]
```diff
@@ -125,29 +163,35 @@ impl<D: Ord> KmerSummarizer<D, Vec<D>> for CountFilterSet<D> {
...
-/// invalid kmers.
+/// BoomHashMap2 Object, check rust-boomphf for details
 #[inline(never)]
 pub fn filter_kmers<K: Kmer, V: Vmer<K>, D1: Clone, DS, S: KmerSummarizer<D1, DS>>(
     seqs: &[(V, Exts, D1)],
     summarizer: S,
     stranded: bool,
-) -> (Vec<(K, (Exts, DS))>, Vec<K>) {
+    report_all_kmers: bool,
+    memory_size: usize,
```
Comment (pmarks) [inline src/filter.rs]: good call on this being configurable.

File: src/filter.rs [hunk 5]
```diff
@@ -95,6 +100,39 @@ impl<D: Ord> KmerSummarizer<D, Vec<D>> for CountFilterSet<D> {
...
+    pub fn new(min_kmer_obs: usize) -> CountFilterSmallInt<D> {
+        CountFilterSmallInt {
+            min_kmer_obs: min_kmer_obs,
+            phantom: PhantomData,
+        }
+    }
+}
+
+
+impl<D: Ord+Debug> KmerSummarizer<D, SmallVec<[D; 4]>> for CountFilterSmallInt<D> {
```
Comment (pmarks) [inline src/filter.rs]: Why does this need to be Debug?

File: src/filter.rs (hunk 3 above)
Comment (k3yavi) [reply src/filter.rs]: I think this bound is required by BoomHashMap2 for the `fmt` trait.

File: src/filter.rs [hunk 6]
```diff
@@ -134,6 +136,65 @@ impl<D: Ord> KmerSummarizer<D, SmallVec<[D; 4]>> for CountFilterSmallInt<D> {
...
+/// Implement this trait to control how multiple observations of a kmer
+/// are carried forward into a DeBruijn graph.
+pub trait MutKmerSummarizer<DI, DO> {
+    /// The input `items` is an iterator over kmer observations. Input observation
+    /// is a tuple of (kmer, extensions, data). The summarize function inspects the
+    /// data and returns a tuple indicating:
+    /// * whether this kmer passes the filtering criteria (e.g. is there a sufficient number of observation)
+    /// * the accumulated Exts of the kmer
+    /// * a summary data object of type `DO` that will be used as a color annotation in the DeBruijn graph.
+    fn summarize<K, F: Iterator<Item = (K, Exts, DI)>>(&mut self, items: F) -> (bool, Exts, DO);
```
Comment (pmarks) [inline src/filter.rs]: nice!  let's just change the original KmerSummarizer trait to be &mut self, rather than introduce a new trait.

Comment (k3yavi) [reply src/filter.rs]: Sounds good, thanks for taking a quick look !!

**Pull request merged**